integrations/clickup/clickup/client.py

```diff
@@ -0,0 +1,172 @@
+            (
+                webhook
+                for webhook in webhooks["webhooks"]
+                if webhook["endpoint"] == webhook_target_app_host
+            ),
+            None,
+        )
+
+        if existing_webhook:
+            logger.info(
+                f"Ocean real time reporting clickup webhook already exists [ID: {existing_webhook['id']}]"
+            )
+            return
+
+        webhook_create_response = await self.client.post(
+            f"{self.clickup_url}/team/{team_id}/webhook",
+            json={
+                "endpoint": webhook_target_app_host,
+                "events": WEBHOOK_EVENTS,
+            },
+        )
+        webhook_create_response.raise_for_status()
+        webhook_create = webhook_create_response.json()
+        logger.info(
+            f"Ocean real time reporting clickup webhook created "
+            f"[ID: {webhook_create['id']}, Team ID: {team_id}]"
+        )
+
+    async def get_paginated_teams(
+        self, params: dict[str, Any] = {}
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        teams_response = await self.client.get(
+            f"{self.clickup_url}/team", params=params
+        )
+        teams_response.raise_for_status()
+        teams = teams_response.json()["teams"]
+        yield teams
+
+    async def _get_paginated_spaces(
+        self, team_id: str, params: dict[str, Any] = {}
+    ) -> list[dict[str, Any]]:
+        spaces_response = await self.client.get(
+            f"{self.clickup_url}/team/{team_id}/space", params=params
+        )
+        spaces_response.raise_for_status()
+        return spaces_response.json()["spaces"]
+
+    async def get_paginated_projects(
+        self, params: dict[str, Any] = {}
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        # getting all teams so as to retrieve the spaces within each team, then using their
+        # ids to get the projects (lists) within each space
+        teams = self.get_paginated_teams(params)
+        async for team_batch in teams:
+            for team in team_batch:
+                spaces = await self._get_paginated_spaces(team["id"], params)
+                projects: list[dict[str, Any]] = []
+
+                for space in spaces:
+                    projects_response = await self.client.get(
+                        f"{self.clickup_url}/space/{space['id']}/list", params=params
+                    )
+                    projects_response.raise_for_status()
+                    # because the port-app-config uses the team ID to relate the projects to the teams
+                    # we add the team ID to each project
+                    # also, the clickup returns the datetime as a timestamp, so we convert it to a
+                    # datetime object
+                    projects.extend(
+                        map(
+                            lambda project: {
+                                **project,
+                                "team_id": team["id"],
```

Comment:
since team_id is an external variable not from the API response, rename it to `__team_id`